### ISSUE TITLE
Replacement of the rendering engine

### ISSUE THREAD

**Issue by volca02** [volca02/openDarkEngine#26]
After trying to force Ogre3d to do stuff the way it's needed, which has proven to be harder on every step (notably light list population for movable objects, lately), it's time to depart ways with it and implement stuff differently. Obvious choice here is not to reinvent wheel when it comes to rendering API abstractions, which should mean easier transitions to new platforms, among other things.

This change will most likely mean shift away from fixed function rendering, as well as getting rid of Ogre's VFS support that was hijacked to be used here. This project does not necessarily need a full blown scene graph implementation, although there has to be support for object attachments - f.ex. camera can be attached to sub-objects of other objects, AI often has keys or valuables on the belt, etc.

Scene management code currently present in scenemanager/ directory will be partially scrapped, partially moved back to worldrep service, that should newly be responsible for static geometry management.

Initial amount of work can be done sideways - loading of BIN (both LGMM and LGMD) files and displaying them can be prototyped in a separate project and backported. The same is probably true for VFS management code, which should live in a fitting service (did original Dark have any service dedicated to VFS?).

**Comment by josephdviviano** [comment]
Has this project ceased development?

Is it the best current starting point for modernizing the dark engine?

**Comment by volca02** [comment]
As far as I am concerned, there simply isn't the time needed to continue this in this moment. I'd have to refactor a lot of code as the choice of OGRE wasn't the best suited and conceptually the repository is quite dated as well - it's not exactly modern C++ at this point.

There is a thin sliver of hope for Dark engine if night dive studios deliver on their promise and opensource the original SS2 source code. Licensing is pretty much the only stopper at this point, the source code for original dark engine can be found on the internet still (won't link for obvious reasons, but at least one copy is on github as we speak).

Edit: You can consider this repository as good source of knowledge maybe. Were I to start working on this again I'd probably build it on top of bgfx with custom rendering code and resource management, but considering the source code for original games is inches away this could be wasted effort, which is frustrating.

**Comment by josephdviviano** [comment]
Thanks - I'll wait to see if they open source SS2 first.

I appreciate your guidance!

**Comment by josephdviviano** [comment]
So I didn't follow your advice to wait. I've been moving pretty fast over here https://github.com/josephdviviano/darkness, building on your excellent work.

I decided however to aim for a hackable / extendable version of the source code (I want to be able to mess with the AI stack), so by the time I'm done, I suspect I'll have diverged from a purely faithful implementation of the original engine.

But since you have all worked so hard on this I thought you would find it interesting.

I still have to do all the hard stuff - audio, AI, physics so maybe I'll get bogged down soon. But anyway, thought I would share.

Hopefully the original source is released by night drive as well, that would be the best of both worlds.

**Comment by hogsy** [comment]
I can't speak for the original authors of openDarkEngine, I'm just an outside observer here and speaking personally, but why did you set up your fork as a fresh repository? It's blatantly based on the code here, but you've scrubbed the original history.

Also, not especially impressed by the Claude usage - it's impossible to know how much of what you're implementing you actually understand. I also think that's going to put people off getting involved, as Claude combined with GPL'd code is a bit of an iffy combination.

**Comment by josephdviviano** [comment]
I've left all references to the codebase intact in the project, including their license and notices throughout. However, the codebase is now so radically refactored already, it was clear to me it would not be ever be integrated back into this project (already, and I've already reimplemented the engine). The commit history was broken due to a mistake on my end that I didn't think was worth resolving since there is no world where this work will be integrated directly back into opde.

the point of my work is not to impress anyone with my knowledge of the game engine's construction. it is to produce an engine with a hackable AI subsystem. as there hasn't been movement on opde project in ~8 years, I didn't expect anyone else would get involved.

I am, however, extremely impressed with how quickly I've been able to get this off the ground. We'll see if I'm able to wrap it up in a reasonable time., and hopefully people who love these games will find it fun.

I'm not clear on why claude + gpl is a iffy combination. I suspect in a few years, it will be the norm.

In any case, open to feedback :) I'm learning just as much as anyone else

**Comment by volca02** [comment]
Hey if you keep this civil (link to original repository, keep the license) I say go for it and experiment.

Honestly looking at the codebase the vibecoding approach will fall short sooner or later, it's not structuring the things all too well. There are shortcomings in the original codebase for sure, f.ex. I'd start with cleaning up the memory ownerships, transition to RAII where possible, get rid of naked pointers in the codebase, get rid of Ogre altogether (including the maths libs - replace those with something independent...). The rendered should be a service, the changes I glimpsed over seemed to go against the design of the project.
There is a lot of research needed to do stuff as it should be done, even from the "replicating original" standpoint. Claude or similar won't likely help unless you are very well informed and hold them accountable for the shortcomings.

**Comment by josephdviviano** [comment]
Thank you! You're right on two fronts, it's very much an experiment on my side (im a scientist, not a game dev) and the refactoring yesterday was already pretty heavy and I haven't gotten that far, so it's very possible that this will fail midway. But for me that's an acceptable outcome as my true goal is to try to learn how to use these tools correctly, which isn't easy to do, but I'm very motivated to try to get it working for this particular project (particularly if I can develop a swappable AI stack - that would be a lot of fun to play with).

The work I've done vs yours, very clear difference in code quality, planning, skill, no doubt :) if I do actually get to my goal, perhaps certain design decisions will actually serve as a good reference for opde, that would be a fun outcome. But I gotta get there first

But yes for pretty much every feature I've needed to correct some elements of Claude's output. I know what the game should be like (I've probably spent more time in the dark engine than any other in my life), but a lot of the under the hood stuff we're figuring out by trial and error and we're probably missing stuff that I'll only uncover when I try to stitch it all together.

**Comment by volca02** [comment]
The code I saw was overstepping the architecture, pulling raw data of some properties, building a huge mesh for the whole mission, etc. - While also doing all this in a spaghetti like manner in a long routines. Not really what I'd call a good direction. It can get you places fast but then you will hit a roadblock wanting to make it more general and systematic.

 It's probably best to start easy, small things, try to understand more than the AI, rather than wanting to push the project much further without understanding what's in place. There already is stuff able to render the missions, but it's using ogre in a hardcoded fassion. Steps could be taken to remove that close connection and replace with a clean rendering interface - I didn't really do a good job separating responsibilities back then. I'd also argue the dynamic typing approach to exposing unknown properties was a good first stab at making the things available, but should be phased out in favor of well implemented structs holding the data.

AI could still be used in a very productive way here, aligning the interfaces with what is known about DE, adding boilerplate implementations of services not yet implemented... But a big effort would have to be put in place of explaining the architecture and expectations. Also putting some tests in was a good idea (I saw some mentions in the commits) - this is what AI can be used to with good results. Again supervision has to be better than the AI itself, the tools can go off rail fast.

**Comment by josephdviviano** [comment]
I agree with everything you said, other than I am not entirely sure if the old workflow of understanding everything and building it correctly the first time is actually the best approach anymore, rather than conjuring spaghetti code and then doing a series of refactor steps. AI lacks vison but does have a substantially larger working memory than us, making "reshaping" the code a lot less expensive.

In any case this is an active experiment and I don't know if this is the right approach, but I appreciate your feedback and expertise. The last comments you gave quickly led to improvements in the codebase

EDIT: I should say that the way I am working is absolutely not collaborator friendly. The merge conflicts would be horrendous, I'm just regularly pushing to main. I recognize this is far from best practice.
